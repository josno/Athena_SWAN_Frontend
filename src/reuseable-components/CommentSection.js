import React, { useState, useRef } from 'react'
import styled from 'styled-components'

import Form from 'react-bootstrap/form'
import Button from 'react-bootstrap/Button'
import { BsPeopleCircle } from 'react-icons/bs'
import { FiThumbsDown, FiThumbsUp } from 'react-icons/fi'

import Utilities from '../Utils/Utilities'

// Function props are passed down from the ArticlePage component
// The logic lives in that component because that is where
// We make updates to the comments and the article
// All logic specific to this component (form) sits in here
// But to actually add and update the comments props we need to send
// it back up to the parent component
const CommentSection = ({ comments, addComment, incrementVote }) => {
  const [form, setForm] = useState({})
  const [errors, setErrors] = useState({})

  // Same as issues page, we use useRef to clearn the form without re-rendering
  const formRef = useRef(null)

  // These are handler function for the form
  // Resets the comment form after a submit
  const handleReset = () => {
    formRef.current.reset()
  }

  // Function to check errors inside submit
  const checkErrors = () => {
    const { name, comment } = form
    const newErrors = {}
    // name errors
    if (!name || name === '') {
      newErrors.name = 'Cannot be blank'
    }

    if (!comment || comment === '') {
      newErrors.comment = 'Cannot be blank'
    }

    return newErrors
  }

  // Updates the state based on input
  const setField = (field, value) => {
    setForm({
      ...form,
      [field]: value,
    })
    // Check and see if errors exist, and remove them from the error object
    if (!!errors[field])
      setErrors({
        ...errors,
        [field]: null,
      })
  }

  const handleSubmit = (e) => {
    e.preventDefault()
    const newErrors = checkErrors()
    if (Object.keys(newErrors).length > 0) {
      // We have errors!
      setErrors(newErrors)
    } else {
      // Calls the function passed down from Article Page
      addComment(form)
      handleReset()
    }
  }

  // Returns the comment form
  const renderForm = () => {
    return (
      <Form
        ref={formRef}
        noValidate
        className="comment-section__form"
        onSubmit={handleSubmit}
      >
        <Form.Group controlId="formGroupName">
          <Form.Label>Name</Form.Label>
          <Form.Control
            placeholder="Name"
            onChange={(e) => setField('name', e.target.value)}
            required
            isInvalid={!!errors.name}
          />
          <Form.Control.Feedback type="invalid">
            {errors.name}
          </Form.Control.Feedback>
        </Form.Group>
        <Form.Group controlId="formGroupTextArea">
          <Form.Label>Write your comments here</Form.Label>
          <Form.Control
            style={{ marginBottom: '0' }}
            as="textarea"
            rows={4}
            onChange={(e) => setField('comment', e.target.value)}
            required
            isInvalid={!!errors.comment}
          />
          <Form.Control.Feedback type="invalid">
            {errors.comment}
          </Form.Control.Feedback>
        </Form.Group>
        <Button
          className="comment-section__button"
          variant="primary"
          type="submit"
        >
          Submit
        </Button>
        <Button
          className="comment-section__button"
          variant="danger"
          onClick={() => handleReset()}
        >
          Cancel
        </Button>
      </Form>
    )
  }

  // Returns all comments
  const renderComments = () => {
    return comments.map((c, index) => (
      <div className="comment-div" key={index}>
        <div className="comment-div__div">
          <h5>
            <span className="comment-div__icon">
              <BsPeopleCircle size={42} />
            </span>
            {c.name}
          </h5>
          <h5>{Utilities.convertDate(c.createdDate)}</h5>
        </div>

        <div>
          <p>{c.comment}</p>
        </div>
        <div className="comment-div__voting">
          <div className="comment-likes__div">
            <span
              className="comment-likes__span "
              onClick={() => incrementVote('dislikes', c.id)}
            >
              <FiThumbsDown className="icon dislike " />
            </span>
            <span className="comment-count">
              {c.dislikes === 0 ? '' : c.dislikes}
            </span>
          </div>
          <div className="comment-likes__div">
            <span
              className="comment-likes__span "
              onClick={() => incrementVote('likes', c.id)}
            >
              <FiThumbsUp className="icon like" />
            </span>
            <span className="comment-count">
              {c.likes === 0 ? '' : c.likes}
            </span>
          </div>
        </div>
      </div>
    ))
  }

  // Here we render the elements for comments and comment form
  // If there are comments, render the comments section, if not do not render
  // If a comment is added, the parent component will get the new data and pass it down

  return (
    <CommentSectionStyles>
      <h2> Article Comments</h2>
      {!comments ? '' : renderComments()}
      {renderForm()}
    </CommentSectionStyles>
  )
}

const CommentSectionStyles = styled.div`
  border-bottom: 1px solid light-gray;
  h5 {
    font-size: 0.9rem;
    font-weight: bold;
  }
  .comment-div {
    border-bottom: 1px solid gray;
    margin-right: 1rem;
  }

  .comment-div__icon {
    margin-right: 1rem;
  }

  .comment-section__form {
    margin-right: 1rem;
  }

  .comment-section__form .form-label {
    margin-top: 2rem;
  }

  .comment-div__div {
    display: flex;
    justify-content: space-between;
    margin-top: 1rem;
  }
  .comment-section__button {
    margin: 1rem;
    float: right;
  }

  .comment-div__voting {
    margin-bottom: 1rem;
    display: flex;
  }

  .comment-likes__div {
    margin-left: 0.5rem;
    width: 3.5rem;
  }

  .comment-likes__span {
    transition: 0.2s;
    :hover {
      cursor: pointer;
    }

    .icon:hover {
      transform: scale(1.1) !important;
    }

    .like:hover {
      color: green;
    }

    .dislike:hover {
      color: red;
    }
  }

  .comment-count {
    margin: 0 0.5rem;
    width: 2rem;
  }
`
export default CommentSection
