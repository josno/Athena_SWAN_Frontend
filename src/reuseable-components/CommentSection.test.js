import { shallow, mount } from 'enzyme'
import CommentSection from './CommentSection'
import toJson from 'enzyme-to-json'

describe('CommentSection', () => {
  it('renders comment section component without crashing', () => {
    const fakeComments = [
      {
        name: 'Test Person',
        comment: 'This is a comment',
        createdDate: {
          seconds: 1621947600,
          nanoseconds: '',
        },
      },
      {
        name: 'Test Person 2',
        comment: 'This is a comment',
        createdDate: {
          seconds: 1621947600,
          nanoseconds: '',
        },
      },
    ]
    shallow(<CommentSection comments={fakeComments} />)
  })
  it('matches the snapshot', () => {
    const fakeComments = [
      {
        name: 'Test Person',
        comment: 'This is a comment',
        createdDate: {
          seconds: 1621947600,
          nanoseconds: '',
        },
      },
      {
        name: 'Test Person 2',
        comment: 'This is a comment',
        createdDate: {
          seconds: 1621947600,
          nanoseconds: '',
        },
      },
    ]
    const wrapper = mount(<CommentSection comments={fakeComments} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
