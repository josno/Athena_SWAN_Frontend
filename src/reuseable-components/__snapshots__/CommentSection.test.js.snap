// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentSection matches the snapshot 1`] = `
<CommentSection
  comments={
    Array [
      Object {
        "comment": "This is a comment",
        "createdDate": Object {
          "nanoseconds": "",
          "seconds": 1621947600,
        },
        "name": "Test Person",
      },
      Object {
        "comment": "This is a comment",
        "createdDate": Object {
          "nanoseconds": "",
          "seconds": 1621947600,
        },
        "name": "Test Person 2",
      },
    ]
  }
>
  <styled.div>
    <div
      className="sc-bdnxRM iSbcTV"
    >
      <h2>
         Article Comments
      </h2>
      <div
        className="comment-div"
        key="0"
      >
        <div
          className="comment-div__div"
        >
          <h5>
            <span
              className="comment-div__icon"
            >
              <BsPeopleCircle
                size={42}
              >
                <IconBase
                  attr={
                    Object {
                      "fill": "currentColor",
                      "viewBox": "0 0 16 16",
                    }
                  }
                  size={42}
                >
                  <svg
                    fill="currentColor"
                    height={42}
                    stroke="currentColor"
                    strokeWidth="0"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 16 16"
                    width={42}
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M13.468 12.37C12.758 11.226 11.195 10 8 10s-4.757 1.225-5.468 2.37A6.987 6.987 0 008 15a6.987 6.987 0 005.468-2.63z"
                      key="0"
                    />
                    <path
                      clipRule="evenodd"
                      d="M8 9a3 3 0 100-6 3 3 0 000 6z"
                      fillRule="evenodd"
                      key="1"
                    />
                    <path
                      clipRule="evenodd"
                      d="M8 1a7 7 0 100 14A7 7 0 008 1zM0 8a8 8 0 1116 0A8 8 0 010 8z"
                      fillRule="evenodd"
                      key="2"
                    />
                  </svg>
                </IconBase>
              </BsPeopleCircle>
            </span>
            Test Person
          </h5>
          <h5>
            Tuesday, May 25, 2021
          </h5>
        </div>
        <div>
          <p>
            This is a comment
          </p>
        </div>
        <div
          className="comment-div__voting"
        >
          <div
            className="comment-likes__div"
          >
            <span
              className="comment-likes__span "
              onClick={[Function]}
            >
              <FiThumbsDown
                className="icon dislike "
              >
                <IconBase
                  attr={
                    Object {
                      "fill": "none",
                      "stroke": "currentColor",
                      "strokeLinecap": "round",
                      "strokeLinejoin": "round",
                      "strokeWidth": "2",
                      "viewBox": "0 0 24 24",
                    }
                  }
                  className="icon dislike "
                >
                  <svg
                    className="icon dislike "
                    fill="none"
                    height="1em"
                    stroke="currentColor"
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    strokeWidth="2"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M10 15v4a3 3 0 0 0 3 3l4-9V2H5.72a2 2 0 0 0-2 1.7l-1.38 9a2 2 0 0 0 2 2.3zm7-13h2.67A2.31 2.31 0 0 1 22 4v7a2.31 2.31 0 0 1-2.33 2H17"
                      key="0"
                    />
                  </svg>
                </IconBase>
              </FiThumbsDown>
            </span>
            <span
              className="comment-count"
            />
          </div>
          <div
            className="comment-likes__div"
          >
            <span
              className="comment-likes__span "
              onClick={[Function]}
            >
              <FiThumbsUp
                className="icon like"
              >
                <IconBase
                  attr={
                    Object {
                      "fill": "none",
                      "stroke": "currentColor",
                      "strokeLinecap": "round",
                      "strokeLinejoin": "round",
                      "strokeWidth": "2",
                      "viewBox": "0 0 24 24",
                    }
                  }
                  className="icon like"
                >
                  <svg
                    className="icon like"
                    fill="none"
                    height="1em"
                    stroke="currentColor"
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    strokeWidth="2"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M14 9V5a3 3 0 0 0-3-3l-4 9v11h11.28a2 2 0 0 0 2-1.7l1.38-9a2 2 0 0 0-2-2.3zM7 22H4a2 2 0 0 1-2-2v-7a2 2 0 0 1 2-2h3"
                      key="0"
                    />
                  </svg>
                </IconBase>
              </FiThumbsUp>
            </span>
            <span
              className="comment-count"
            />
          </div>
        </div>
      </div>
      <div
        className="comment-div"
        key="1"
      >
        <div
          className="comment-div__div"
        >
          <h5>
            <span
              className="comment-div__icon"
            >
              <BsPeopleCircle
                size={42}
              >
                <IconBase
                  attr={
                    Object {
                      "fill": "currentColor",
                      "viewBox": "0 0 16 16",
                    }
                  }
                  size={42}
                >
                  <svg
                    fill="currentColor"
                    height={42}
                    stroke="currentColor"
                    strokeWidth="0"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 16 16"
                    width={42}
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M13.468 12.37C12.758 11.226 11.195 10 8 10s-4.757 1.225-5.468 2.37A6.987 6.987 0 008 15a6.987 6.987 0 005.468-2.63z"
                      key="0"
                    />
                    <path
                      clipRule="evenodd"
                      d="M8 9a3 3 0 100-6 3 3 0 000 6z"
                      fillRule="evenodd"
                      key="1"
                    />
                    <path
                      clipRule="evenodd"
                      d="M8 1a7 7 0 100 14A7 7 0 008 1zM0 8a8 8 0 1116 0A8 8 0 010 8z"
                      fillRule="evenodd"
                      key="2"
                    />
                  </svg>
                </IconBase>
              </BsPeopleCircle>
            </span>
            Test Person 2
          </h5>
          <h5>
            Tuesday, May 25, 2021
          </h5>
        </div>
        <div>
          <p>
            This is a comment
          </p>
        </div>
        <div
          className="comment-div__voting"
        >
          <div
            className="comment-likes__div"
          >
            <span
              className="comment-likes__span "
              onClick={[Function]}
            >
              <FiThumbsDown
                className="icon dislike "
              >
                <IconBase
                  attr={
                    Object {
                      "fill": "none",
                      "stroke": "currentColor",
                      "strokeLinecap": "round",
                      "strokeLinejoin": "round",
                      "strokeWidth": "2",
                      "viewBox": "0 0 24 24",
                    }
                  }
                  className="icon dislike "
                >
                  <svg
                    className="icon dislike "
                    fill="none"
                    height="1em"
                    stroke="currentColor"
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    strokeWidth="2"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M10 15v4a3 3 0 0 0 3 3l4-9V2H5.72a2 2 0 0 0-2 1.7l-1.38 9a2 2 0 0 0 2 2.3zm7-13h2.67A2.31 2.31 0 0 1 22 4v7a2.31 2.31 0 0 1-2.33 2H17"
                      key="0"
                    />
                  </svg>
                </IconBase>
              </FiThumbsDown>
            </span>
            <span
              className="comment-count"
            />
          </div>
          <div
            className="comment-likes__div"
          >
            <span
              className="comment-likes__span "
              onClick={[Function]}
            >
              <FiThumbsUp
                className="icon like"
              >
                <IconBase
                  attr={
                    Object {
                      "fill": "none",
                      "stroke": "currentColor",
                      "strokeLinecap": "round",
                      "strokeLinejoin": "round",
                      "strokeWidth": "2",
                      "viewBox": "0 0 24 24",
                    }
                  }
                  className="icon like"
                >
                  <svg
                    className="icon like"
                    fill="none"
                    height="1em"
                    stroke="currentColor"
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    strokeWidth="2"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M14 9V5a3 3 0 0 0-3-3l-4 9v11h11.28a2 2 0 0 0 2-1.7l1.38-9a2 2 0 0 0-2-2.3zM7 22H4a2 2 0 0 1-2-2v-7a2 2 0 0 1 2-2h3"
                      key="0"
                    />
                  </svg>
                </IconBase>
              </FiThumbsUp>
            </span>
            <span
              className="comment-count"
            />
          </div>
        </div>
      </div>
      <Form
        className="comment-section__form"
        inline={false}
        noValidate={true}
        onSubmit={[Function]}
      >
        <form
          className="comment-section__form"
          noValidate={true}
          onSubmit={[Function]}
        >
          <FormGroup
            controlId="formGroupName"
          >
            <div
              className="form-group"
            >
              <FormLabel
                column={false}
                srOnly={false}
              >
                <label
                  className="form-label"
                  htmlFor="formGroupName"
                >
                  Name
                </label>
              </FormLabel>
              <FormControl
                isInvalid={false}
                onChange={[Function]}
                placeholder="Name"
                required={true}
              >
                <input
                  className="form-control"
                  id="formGroupName"
                  onChange={[Function]}
                  placeholder="Name"
                  required={true}
                />
              </FormControl>
              <Feedback
                type="invalid"
              >
                <div
                  className="invalid-feedback"
                />
              </Feedback>
            </div>
          </FormGroup>
          <FormGroup
            controlId="formGroupTextArea"
          >
            <div
              className="form-group"
            >
              <FormLabel
                column={false}
                srOnly={false}
              >
                <label
                  className="form-label"
                  htmlFor="formGroupTextArea"
                >
                  Write your comments here
                </label>
              </FormLabel>
              <FormControl
                as="textarea"
                isInvalid={false}
                onChange={[Function]}
                required={true}
                rows={4}
                style={
                  Object {
                    "marginBottom": "0",
                  }
                }
              >
                <textarea
                  className="form-control"
                  id="formGroupTextArea"
                  onChange={[Function]}
                  required={true}
                  rows={4}
                  style={
                    Object {
                      "marginBottom": "0",
                    }
                  }
                />
              </FormControl>
              <Feedback
                type="invalid"
              >
                <div
                  className="invalid-feedback"
                />
              </Feedback>
            </div>
          </FormGroup>
          <Button
            active={false}
            className="comment-section__button"
            disabled={false}
            type="submit"
            variant="primary"
          >
            <button
              className="comment-section__button btn btn-primary"
              disabled={false}
              type="submit"
            >
              Submit
            </button>
          </Button>
          <Button
            active={false}
            className="comment-section__button"
            disabled={false}
            onClick={[Function]}
            variant="danger"
          >
            <button
              className="comment-section__button btn btn-danger"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Cancel
            </button>
          </Button>
        </form>
      </Form>
    </div>
  </styled.div>
</CommentSection>
`;
